test/src/test/java/org/corfudb/integration/LogReplicationListenerIT.java

```diff
@@ -0,0 +1,667 @@
+        initializeCorfu();
+        openTables();
+
+        // Start snapshot sync if requested
+        if (startInSnapshotSync) {
+            writeToStatusTable(false);
+        }
+
+        final int numUpdates = 10;
+        CountDownLatch countDownLatch = new CountDownLatch(numUpdates);
+        lrListener = new LRTestListener(store, namespace, userTag, Arrays.asList(userTableName), countDownLatch);
+
+        // Subscribe the listener
+        store.subscribeLogReplicationListener(lrListener, namespace, userTag, Arrays.asList(userTableName));
+
+        // End snapshot sync if it had been requested
+        if (startInSnapshotSync) {
+            writeToStatusTable(true);
+        }
+
+        // Write numUpdates records in the data table
+        writeToDataTable(numUpdates, 0);
+
+        // Wait for the data to arrive.  Since subscription is asynchronous, streaming updates may be lesser than
+        // numUpdates.  But the final number of entries seen by mergeTableOnSubscription() + streaming updates must
+        // be equal to numUpdates
+        countDownLatch.await();
+
+        // If the test started when in log entry sync, performFullSync() should not find any existing data as no
+        // updates have been written prior to subscription.
+        if (!startInSnapshotSync) {
+            Assert.assertTrue(lrListener.getExistingEntries().isEmpty());
+        }
+
+        // Verify the sequence(timestamp) of the streaming updates
+        verifyUpdatesSequence(lrListener.getUpdates());
+
+        // Verify all the data observed by the listener - existing entries before subscription + updates afterwards
+        verifyData(lrListener.getUpdates(), lrListener.getExistingEntries());
+    }
+
+    @Test
+    public void testMultipleWritesToDataTableInTxStartInLogEntrySync() throws Exception {
+        testMultipleWritesToDataTableInTx(false);
+    }
+
+    @Test
+    public void testMultipleWritesToDataTableInTxStartInSnapshotSync() throws Exception {
+        testMultipleWritesToDataTableInTx(true);
+    }
+
+    /**
+     * This test verifies that all data written to the user table is received in increasing order of the timestamps and
+     * the streaming updates match the data in the table.  Multiple entries to the table are written in a single
+     * transaction (batch size >1).
+     * The final number of entries seen by performFullSync(if starting in snapshot sync) + streaming updates must be
+     * equal to the total number of writes to the table.
+     */
+    private void testMultipleWritesToDataTableInTx(boolean startInSnapshotSync) throws Exception {
+        initializeCorfu();
+        openTables();
+
+        // Start snapshot sync if requested
+        if (startInSnapshotSync) {
+            writeToStatusTable(false);
+        }
+
+        final int numUpdates = 10;
+        final int numStreamingUpdates = 1;
+        CountDownLatch countDownLatch = new CountDownLatch(numStreamingUpdates);
```

Comment:
I thought we have similar issue for differentiating log entry & snapshot sync count down latch size.

Also, another issue is I think this test is for verifying `onNext()` is only executed once as we apply all the data in a single txn right? but actually the `countDownLatch` in LRTestListener is counted down per entry?

not sure is it easy to expose for this listener, but seems like we need a txnLatch for the purpose of this test?